File: workflows/src/main/java/net/corda/solarsystem/flows/LaunchProbeFlow.java
package net.corda.solarsystem.flows;

import net.corda.solarsystem.contracts.ProbeContract;
import net.corda.solarsystem.states.ProbeState;
import net.corda.systemflows.CollectSignaturesFlow;
import net.corda.systemflows.FinalityFlow;
import net.corda.systemflows.ReceiveFinalityFlow;
import net.corda.systemflows.SignTransactionFlow;
import net.corda.v5.application.flows.*;
import net.corda.v5.application.flows.flowservices.FlowEngine;
import net.corda.v5.application.flows.flowservices.FlowIdentity;
import net.corda.v5.application.flows.flowservices.FlowMessaging;
import net.corda.v5.application.identity.AbstractParty;
import net.corda.v5.application.identity.CordaX500Name;
import net.corda.v5.application.identity.Party;
import net.corda.v5.application.injection.CordaInject;
import net.corda.v5.application.services.IdentityService;
import net.corda.v5.application.services.json.JsonMarshallingService;
import net.corda.v5.base.annotations.Suspendable;
import net.corda.v5.ledger.UniqueIdentifier;
import net.corda.v5.ledger.contracts.Command;
import net.corda.v5.ledger.services.NotaryLookupService;
import net.corda.v5.ledger.services.TransactionService;
import net.corda.v5.ledger.transactions.SignedTransaction;
import net.corda.v5.ledger.transactions.SignedTransactionDigest;
import net.corda.v5.ledger.transactions.TransactionBuilder;
import net.corda.v5.ledger.transactions.TransactionBuilderFactory;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.stream.Collectors;

@InitiatingFlow
@StartableByRPC
class LaunchProbeFlowJava implements Flow<SignedTransactionDigest> {

    private final RpcStartFlowRequestParameters params;

    @CordaInject
    private FlowEngine flowEngine;

    @CordaInject
    private FlowIdentity flowIdentity;

    @CordaInject
    private FlowMessaging flowMessaging;

    @CordaInject
    private TransactionService transactionService;

    @CordaInject
    private TransactionBuilderFactory transactionBuilderFactory;

    @CordaInject
    private IdentityService identityService;

    @CordaInject
    private NotaryLookupService notaryLookup;

    @CordaInject
    private JsonMarshallingService jsonMarshallingService;

    @JsonConstructor
    public LaunchProbeFlowJava(RpcStartFlowRequestParameters params) {
        this.params = params;
    }

    @Override
    @Suspendable
    public SignedTransactionDigest call() {

        Map<String, String> parametersMap = jsonMarshallingService.parseJson(params.getParametersInJson(), Map.class);

        String message;
        boolean planetaryOnly;
        CordaX500Name target;

        if (!parametersMap.containsKey("message"))
            throw new BadRpcStartFlowRequestException("Parameter \"message\" missing.");
        else
            message = parametersMap.get("message");

        if (!parametersMap.containsKey("planetaryOnly"))
            throw new BadRpcStartFlowRequestException("Parameter \"planetaryOnly\" missing.");
        else
            planetaryOnly = Boolean.getBoolean(parametersMap.get("planetaryOnly"));

        if (!parametersMap.containsKey("target"))
            throw new BadRpcStartFlowRequestException("Parameter \"target\" missing.");
        else
            target = CordaX500Name.parse(parametersMap.get("target"));

        Party sender = flowIdentity.getOurIdentity();
        Party recipientParty = identityService.partyFromName(target);
        if (recipientParty == null) throw new NoSuchElementException("No party found for X500 name $target");
        Party notary = notaryLookup.getNotaryIdentities().get(0);

        // Stage 1.
        // Generate an unsigned transaction.
        ProbeState probeState = new ProbeState(message, planetaryOnly, flowIdentity.getOurIdentity(), recipientParty, new UniqueIdentifier());
        Command txCommand = new Command(new ProbeContract.Commands.Launch(), probeState.getParticipants().stream().map(AbstractParty::getOwningKey).collect(Collectors.toList()));

        TransactionBuilder txBuilder = transactionBuilderFactory.create()
                .setNotary(notary)
                .addOutputState(probeState, ProbeContract.getID())
                .addCommand(txCommand);

        // Stage 2.
        // Verify that the transaction is valid.
        txBuilder.verify();

        // Stage 3.
        // Sign the transaction.
        SignedTransaction partSignedTx = txBuilder.sign();

        // Stage 4.
        // Send the state to the counterparty, and receive it back with their signature.
        FlowSession otherPartySession = flowMessaging.initiateFlow(recipientParty);

        SignedTransaction fullySignedTx = flowEngine.subFlow(
                new CollectSignaturesFlow(
                        partSignedTx,
                        Arrays.asList(otherPartySession)
                )
        );

        // Stage 5.
        // Notarise and record the transaction in both parties' vaults.
        SignedTransaction notarisedTx = flowEngine.subFlow(
                new FinalityFlow(fullySignedTx, Arrays.asList(otherPartySession))
        );

        //Step 6.
        // Return Json output
        return new SignedTransactionDigest(
                notarisedTx.getId(),
                Collections.singletonList(jsonMarshallingService.formatJson(notarisedTx.getTx().getOutputStates().get(0))),
                notarisedTx.getSigs()
        );

    }
}

@InitiatedBy(LaunchProbeFlowJava.class)
class LaunchProbeFlowAcceptorJava implements Flow<SignedTransaction> {

    private final FlowSession counterPartySession;

    @CordaInject
    private FlowEngine flowEngine;

    LaunchProbeFlowAcceptorJava(FlowSession counterPartySession) {
        this.counterPartySession = counterPartySession;
    }

    @Override
    @Suspendable
    public SignedTransaction call() {
        SignedTransaction signedTransaction = flowEngine.subFlow(new MySignTransactionFlow(counterPartySession));

        return flowEngine.subFlow(new ReceiveFinalityFlow(counterPartySession, signedTransaction.getId()));
    }

    public static class MySignTransactionFlow extends SignTransactionFlow {
        MySignTransactionFlow(FlowSession counterpartySession) {
            super(counterpartySession);
        }

        @Override
        protected void checkTransaction(@NotNull SignedTransaction stx) {
        }
    }
}

File: workflows/src/main/kotlin/net/corda/solarsystem/flows/LaunchProbeFlow.kt
package net.corda.solarsystem.flows

import net.corda.solarsystem.contracts.ProbeContract
import net.corda.solarsystem.states.ProbeState
import net.corda.systemflows.CollectSignaturesFlow
import net.corda.systemflows.FinalityFlow
import net.corda.systemflows.ReceiveFinalityFlow
import net.corda.systemflows.SignTransactionFlow
import net.corda.v5.application.flows.*
import net.corda.v5.application.flows.flowservices.FlowEngine
import net.corda.v5.application.flows.flowservices.FlowIdentity
import net.corda.v5.application.flows.flowservices.FlowMessaging
import net.corda.v5.application.identity.CordaX500Name
import net.corda.v5.application.injection.CordaInject
import net.corda.v5.application.services.IdentityService
import net.corda.v5.application.services.json.JsonMarshallingService
import net.corda.v5.application.services.json.parseJson
import net.corda.v5.base.annotations.Suspendable
import net.corda.v5.ledger.contracts.Command
import net.corda.v5.ledger.contracts.requireThat
import net.corda.v5.ledger.services.NotaryLookupService
import net.corda.v5.ledger.services.TransactionService
import net.corda.v5.ledger.transactions.SignedTransaction
import net.corda.v5.ledger.transactions.SignedTransactionDigest
import net.corda.v5.ledger.transactions.TransactionBuilderFactory

/**
 * This flow allows two parties (the [Launcher] and the [Target]) to say hello to one another via the [ProbeState].
 *
 * In our simple example, the [Acceptor] will only accepts a valid Probe.
 *
 * These flows have deliberately been implemented by using only the call() method for ease of understanding. In
 * practice, we would recommend splitting up the various stages of the flow into sub-routines.
 *
 * All methods called within the [FlowLogic] sub-class need to be annotated with the @Suspendable annotation.
 */
@InitiatingFlow
@StartableByRPC
class LaunchProbeFlow @JsonConstructor constructor(private val params: RpcStartFlowRequestParameters) :
    Flow<SignedTransactionDigest> {
    @CordaInject
    lateinit var flowEngine: FlowEngine
    @CordaInject
    lateinit var flowIdentity: FlowIdentity
    @CordaInject
    lateinit var flowMessaging: FlowMessaging
    @CordaInject
    lateinit var transactionService: TransactionService
    @CordaInject
    lateinit var transactionBuilderFactory: TransactionBuilderFactory
    @CordaInject
    lateinit var identityService: IdentityService
    @CordaInject
    lateinit var notaryLookup: NotaryLookupService
    @CordaInject
    lateinit var jsonMarshallingService: JsonMarshallingService

    /**
     * The flow logic is encapsulated within the call() method.
     */
    @Suspendable
    override fun call(): SignedTransactionDigest {
        // parse parameters
        val mapOfParams: Map<String, String> = jsonMarshallingService.parseJson(params.parametersInJson)

        val message = with(mapOfParams["message"] ?: throw BadRpcStartFlowRequestException("Parameter \"message\" missing.")) {
            this
        }

        val planetaryOnly = with(mapOfParams["planetaryOnly"] ?: throw BadRpcStartFlowRequestException("Parameter \"planetaryOnly\" missing.")) {
            this.toBoolean()
        }

        val target = with(mapOfParams["target"] ?: throw BadRpcStartFlowRequestException("Parameter \"target\" missing.")) {
            CordaX500Name.parse(this)
        }
        val recipientParty = identityService.partyFromName(target)
            ?: throw NoSuchElementException("No party found for X500 name $target")

        val notary = notaryLookup.notaryIdentities.first()

        // Stage 1.
        // Generate an unsigned transaction.
        val probeState = ProbeState(message, planetaryOnly, flowIdentity.ourIdentity, recipientParty)
        val txCommand = Command(ProbeContract.Commands.Launch(), probeState.participants.map { it.owningKey })
        val txBuilder = transactionBuilderFactory.create()
            .setNotary(notary)
            .addOutputState(probeState, ProbeContract.ID)
            .addCommand(txCommand)

        // Stage 2.
        // Verify that the transaction is valid.
        txBuilder.verify()

        // Stage 3.
        // Sign the transaction.
        val partSignedTx = txBuilder.sign()

        // Stage 4.
        // Send the state to the counterparty, and receive it back with their signature.
        val otherPartySession = flowMessaging.initiateFlow(recipientParty)
        val fullySignedTx = flowEngine.subFlow(
            CollectSignaturesFlow(
                partSignedTx, setOf(otherPartySession),
            )
        )

        // Stage 5.
        // Notarise and record the transaction in both parties' vaults.
        val notarisedTx = flowEngine.subFlow(
            FinalityFlow(
                fullySignedTx, setOf(otherPartySession),
            )
        )

        return SignedTransactionDigest(
            notarisedTx.id,
            notarisedTx.tx.outputStates.map { output -> jsonMarshallingService.formatJson(output) },
            notarisedTx.sigs
        )
    }
}

@InitiatedBy(LaunchProbeFlow::class)
class LaunchProbeFlowAcceptor(val otherPartySession: FlowSession) : Flow<SignedTransaction> {
    @CordaInject
    lateinit var flowEngine: FlowEngine

    // instead, for now, doing this so it can be unit tested separately:
    fun isValid(stx: SignedTransaction) {
        requireThat {

            val output = stx.tx.outputs.single().data
            "This must be an Probe transaction." using (output is ProbeState)
        }
    }

    @Suspendable
    override fun call(): SignedTransaction {
        val signTransactionFlow = object : SignTransactionFlow(otherPartySession) {
            override fun checkTransaction(stx: SignedTransaction) = isValid(stx)
        }
        val txId = flowEngine.subFlow(signTransactionFlow).id
        return flowEngine.subFlow(ReceiveFinalityFlow(otherPartySession, expectedTxId = txId))
    }
}


